src/environment.js

```diff
@@ -0,0 +1,1114 @@
+            }
+        }
+    },
+
+    async translations(path) {
+        var url = util.apiURL(
+            (path ? path : this.__kumascript.env.url) + '$json'
+        );
+        var json = await this.__kumascript.MDN.fetchJSONResource(url);
+        var result = [];
+        if (json != null) {
+            result = json.translations || [];
+        }
+        return result;
+    }
+};
+
+class Environment {
+    // Intialize an environment object that will be used to render
+    // all of the macros in one document or page. We pass in a context
+    // object (which may come from HTTP request headers) that gives
+    // details like the page title and URL. These are available to macros
+    // through the global 'env' object, and some of the properties
+    // are also copied onto the global 'page' object.
+    //
+    // Note that we don't use the Environment object directly when
+    // executing macros. Instead call getExecutionContext(), supplying
+    // the macro arguments list to get an object specific for executing
+    // one macro.
+    //
+    // Note that we pass the Templates object when we create an Environment.
+    // this is so that macros can recursively execute other named macros
+    // in the same environment.
+    //
+    // The optional third argument is for use only by tests. Setting it to
+    // true makes us not freeze the environment so that tests can stub out
+    // methods in the API like mdn.fetchJSONResources
+    //
+    constructor(perPageContext, templates, testing = false) {
+        // Freeze an object unless we're in testing mode
+        function freeze(o) {
+            return testing ? o : Object.freeze(o);
+        }
+
+        /**
+         * For each function-valued property in o, bind that function
+         * to the specified bindings object, if there is one. Also,
+         * create lowercase and titlecase variants of each property in
+         * o, and finally, freeze the object o and return it.
+         *
+         * The binding means that the functions defined in this file
+         * can use `this.__kumascript` to refer to the global
+         * kumascript environment so that they can do
+         * `this.__kumascript.env.locale` for example.
+         *
+         * The case-insensitive variants implement legacy behavior in
+         * KumaScript, where macros can use case-insensitive names of
+         * objects and methods.
+         *
+         * And the freeze() call is a safety measure to prevent
+         * macros from modifying the execution environment.
+         */
+        function prepareProto(o, binding) {
+            let p = {};
+            for (let [key, value] of Object.entries(o)) {
+                if (binding && typeof value === 'function') {
+                    value = value.bind(binding);
+                }
+                p[key] = value;
+                p[key.toLowerCase()] = value;
+                p[key[0].toUpperCase() + key.slice(1)] = value;
```

Comment:
It could be implemented with a Proxy. I'm not sure that I agree that it should be, however. It is not like we're planning to broaden the case-insensitive API surface. Rather than doing proxies, I'd rather go find all the non-canonical uses of these APIs and convert them to be canonical.

Again, however, not as part of this patch.